### ISSUE TITLE
Using outdated counts when running counter_culture_fix_counts?

### ISSUE THREAD

**Issue by magnusvk** [magnusvk/counter_culture#24]
`counter_culture_fix_counts` first loads all the correct counts into memory, then iterates over all instances and compares counts. Now, if in between loading the correct counts and comparing them in the loop, new children were created or destroyed, this count will be off and reset to an incorrect, outdated value.

Maybe we can use an update_all with a JOIN to let the database handle this atomically? Might this also be faster?

We could also put the fetching of each correct count and the corresponding update into a transaction, but then that would slow the whole process way down.

**Comment by xtagon** [comment]
Would changing it to do an update_all with a JOIN help out with https://github.com/magnusvk/counter_culture/issues/79 at all?

Even though that issue was closed, it is actually still a pressing issue for me and I haven't found a suitible solution yet.

**Comment by magnusvk** [comment]
@xtagon I'm not sure what you mean to be honest.

**Comment by xtagon** [comment]
> Maybe we can use an update_all with a JOIN to let the database handle this atomically? Might this also be faster?

@magnusvk I was hoping this meant `counter_culture_fix_counts` would be faster in that case

**Comment by magnusvk** [comment]
@xtagon it uses `update_all`: https://github.com/magnusvk/counter_culture/blob/master/lib/counter_culture/reconciler.rb#L118

**Comment by xtagon** [comment]
@magnusvk Yes it does, but it does 1 `update_all` per primary key, so it's effectively a singular update.

In the end I was able to do a single `update_all` to compute counts for a table of millions of records in under a minute, but my use case is very simplified compared to what counter_culture supports so I can see how that would be hard to add to the library (with things like soft deletion, advanced associations, etc.)

**Comment by magnusvk** [comment]
Great, glad you could get it resolved. :)
